libs/core/algorithms/include/hpx/parallel/algorithms/fold.hpp

```diff
@@ -18,57 +22,156 @@ namespace hpx::parallel { namespace detail {
         HPX_HOST_DEVICE static constexpr T sequential(
             ExPolicy&&, FwdIter first, Sent last, T&& init, F&& f)
         {
-            hpx::reduce(first, last, HPX_FORWARD(T, init), HPX_FORWARD(F, f));
+            if (first == last)
+                return HPX_MOVE(init);
+            T acc = HPX_MOVE(init);
+            while (first != last)
+                acc = HPX_MOVE(f(HPX_MOVE(acc), *first++));
+            return HPX_MOVE(acc);
```

Comment:
There is no need to move returned values from functions, those are rvalues to begin with (e.g. writing things like `a = HPX_MOVE(f());`. 

There is usually also no need to move the values in the return statement. Those are treated as rvalues as well. writing `return HPX_MOVE(bar);` does more harm than good as it disables RTVO.

But in general, what is your rationale for reimplementing the sequential fold? We do have that already somewhere for implementing the sequential `reduce()`.

Reply:
> There is no need to move returned values from functions, those are rvalues to begin with (e.g. writing things like a = HPX_MOVE(f());.

sure will change it.


> There is usually also no need to move the values in the return statement. Those are treated as rvalues as well. writing return HPX_MOVE(bar); does more harm than good as it disables RTVO.
The [possible implementation](https://en.cppreference.com/w/cpp/algorithm/ranges/fold_left) in the standard seems to return moved values


> But in general, what is your rationale for reimplementing the sequential fold? We do have that already somewhere for implementing the sequential reduce().
I can change the implementation to hpx::reduce if that's preferred.

Reply:
> I can change the implementation to hpx::reduce if that's preferred.

I think you could directly call our sequential_reduce implementation.

Reply:
To use the fold algorithm, the user needs to `#include <hpx/algorithm/parallel/fold.hpp>`, should we force them to include another entire .hpp file to skip writing one loop?

Reply:
I'm not sure what you mean by that. Could you elaborate, please?